rfd/xxxx-gcp-server-discovery.md

```diff
@@ -0,0 +1,285 @@
+    #!/bin/sh
+    curl https://.../teleport-pubkey.asc ...
+    echo "deb [signed-by=... stable main" | tee ... > /dev/null
+    apt-get update
+    apt-get install teleport
+    teleport node configure --auth-agent=... --join-method=gcp --token-name=gcp-token
+  # Any resource in Teleport can automatically expire.
+  expires: 0001-01-01T00:00:00Z
+```
+
+Unless overridden by a user, a default teleport installer command will be
+generated that is appropriate for the current running version and operating
+system initially supporting DEB and RPM based distros that Teleport already
+provides packages for.
+
+To run commands on a VM, the Discovery Service will create a short-lived
+ssh key pair and add the public key to the VM via its metadata. Then it will
+run the installer on the VM over SSH.
+
+> Note: GCP VMs using [OS Login](https://cloud.google.com/compute/docs/oslogin) do not support SSH keys in instance metadata.
+
+The Discovery Service's service account will require the following permissions:
+
+- `compute.instances.setMetadata`
+- `compute.instances.get`
+- `compute.instances.list`
+
+### GCP join method
+
+In order to register GCP virtual machines, a new `gcp` join method will be
+created. The `gcp` join method will be
+an oidc-based join method like `github`, `circleci`, etc. The token will be fetched from the VM's
+instance metadata at the internal URL
+`http://metadata/computeMetadata/v1/instance/service-accounts/default/identity?audience=AUDIENCE&format=FORMAT&licenses=LICENSES`,
+with the audience claim set to the name of the Teleport cluster. The issuer
+will always be `https://accounts.google.com`. The rest of the registration
+process will be identical to that of the
+[other oidc join methods](https://github.com/gravitational/teleport/blob/master/rfd/0079-oidc-joining.md#auth-server-support).
+
+The joining VM will need a [service account](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances)
+assigned to it to be able to generate id tokens. No permissions on the account
+are needed.
+
+#### Teleport Configuration
+
+The existing provision token type can be extended to support GCP
+authentication, using new GCP-specific fields in the token rules section.
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example_gcp_token
+spec:
+  roles: [Node, Kube, Db]
+  gcp:
+    allow:
+      # IDs of projects from which nodes can join. At least one required.
+      - project_ids: ['p1', 'p2']
+        # Regions and/or zones from which nodes can join. If empty or omitted,
+        # nodes from any location are allowed.
+        locations: ['l1', 'l2']
+        # Emails associated with service accounts assigned to joining nodes. If
+        # empty or omitted, nodes with any email are allowed.
+        emails: ['e1@example.com', 'e2@example.com']
+        # Names of joining instances. If empty or omitted, nodes with any name
+        # are allowed.
+        instance_names: ['i1', 'i2']
```

Comment:
This tbh doesn't sounds like it's too useful to have. Why would anyone use auto-discovery if they only wanted to limit to specific instances?

I'd cut this from the scope.